### ISSUE TITLE
Implement shade constructor and class system

### ISSUE THREAD

**Issue by njtierney** [njtierney/naniar#190]
Building upon #189

# shade

A shade is built on top of factor, and has, at the minimum, the levels "!NA" and "NA". It also has the capability to contain other levels of missings, which start with "NA_". For example, NA_broken

# Current progress

Currently `shade` falls over with a strange bug where anything not in the first position in a vector doesn't get the special labels.

For example:

``` r
library(naniar)

sh_1 <- shade(c(3, 1, 2, NA), broken = 3)

sh_1
#> [1] NA_broken !NA       !NA       NA       
#> Levels: NA !NA NA_broken
paste0(sh_1)
#> [1] "NA_broken" "!NA"       "!NA"       "NA"


sh_2 <- shade(c(3, 1, 2, NA), broken = 1)
sh_2
#> [1] !NA  <NA> !NA  NA  
#> Levels: NA !NA NA_broken
paste0(sh_2)
#> [1] "!NA" "NA"  "!NA" "NA"

sh_3 <- shade(c(3, 1, 2, NA), broken = 2)
sh_3
#> [1] !NA  !NA  <NA> NA  
#> Levels: NA !NA NA_broken
paste0(sh_3)
#> [1] "!NA" "!NA" "NA"  "NA"
```

Created on 2018-08-12 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by njtierney** [comment]
Going on from the [Advanced R book chapter on S3 methods](https://adv-r.hadley.nz/s3.html#constructors), I should have:

- Constructor methods (`new_shade`)
- validators (`validate_shade`) - is it a factor, does it contain at least the levels `!NA` and `NA`, and do additional levels have more than `NA_` 
- helpers (`shade`, and `modify_shade`)
- object style: This is a **vector style** of object

What should I expect `shade` to do?

- Convert a vector into missing or not missing: `NA` and `!NA`, respectively.
- Possibly: Have a way to create new values and levels of missing `NA_{suffix}`.

What should I expect `modify_shade` to do?

- This function should add levels to the shade, and possibly modify in place - although this might not work at the vector-level, as modifying the levels could be dependent upon values in other vectors (e.g., `a_NA` could have a special missing values that is based on values in `b` or `c`).

**Comment by njtierney** [comment]
Also a note that I will need to revisit the functions:

- `is_shadow` 
- `are_shadow`, and co

as these are currently written using `grepl` to find the columns that have "_NA".

**Comment by njtierney** [comment]
shade has been implemented in the special-missings branch, and so this issue should be closed soon

**Comment by njtierney** [comment]
closed by #202
